pyproject.toml

```diff
@@ -36,6 +36,7 @@ dependencies = [
     "bashlex",
     "gitpython",
     "packaging",
+    "rapids-metadata>=0.2.0,<0.3.0.dev0",
```

Comment:
```suggestion
    "rapids-metadata>=0.2.0"
```

In the README for `rapids-metadata` ([link](https://github.com/rapidsai/rapids-metadata)), I see:

> *this package acts as an unpinned dependency of `pre-commit-hooks`*

And I recall that being one of the main motivation for creating a `rapids-metadata` package... to be able to add new project metadata and have `pre-commit` hooks start recognizing it without needing to do a round of updating the hook version across repos.

Given that, I think this should be a **floor** but not have a ceiling.

---

That's also making me think though... is using an unpinned dependency going to work the way we want it to given how `pre-commit` caches packages? I don't think `pre-commit` will update the virtualenv(s) it uses for hooks until the hook versions themselves change.

See, for example, https://github.com/pre-commit/pre-commit/issues/2633.

That'd mean a run of `pre-commit run` in an environment that already has caches set up wouldn't automatically use the newest `rapids-metadata`.

Some such environments:

* anywhere you do local, interactive development
* CI using a mounted-in cache (https://pre-commit.com/#managing-ci-caches)

I think this problem deserves some design work.

Some options I can think of (in no order)

* force the `pre-commit` cache used in RAPIDS CI to be recreated regularly
   - * the `actions/cache` action doesn't explicitly have a time-to-live type of setting ([docs](https://github.com/actions/cache?tab=readme-ov-file#inputs))
   - *but we could achieve that by adding a date component to the key here: RAPIDS: https://github.com/rapidsai/shared-workflows/blob/ecb33372bbe2f9e33502b9472b0b1d0651848657/.github/workflows/checks.yaml#L83-L86*
* host the `rapids-metadata` data behind a network call, and have the hook download it every time it runs
  - *like https://version.gpuci.io/rapids/24.08 for mapping from UCX to RAPIDS versions*
  - *or even just as a big JSON file that's downloaded from some remote object storage*
* set up some automation that runs `pre-commit autoupdate` on a regular schedule across all RAPIDS repos and submits PRs
  - *can't find the link now but I know @bdice has mentioned this in the past*

Reply:
> Given that, I think this should be a floor but not have a ceiling.

I disagree. Once we publish 1.0, breaking API changes will trigger a major version update. We don't want clients that expect the 1.0 API to suddenly use 2.0.

> is using an unpinned dependency going to work the way we want it to given how pre-commit caches packages? I don't think pre-commit will update the virtualenv(s) it uses for hooks until the hook versions themselves change.

Ouch... you're right. I hadn't thought of that.

> host the rapids-metadata data behind a network call, and have the hook download it every time it runs

I like this idea best. It automatically covers all possible use cases.

Reply:
> *Once we publish 1.0, breaking API changes will trigger a major version update. We don't want clients that expect the 1.0 API to suddenly use 2.0.*

As a general approach to distributing libraries I agree with you, but here we're using one version number to convey changes to 2 different things:

* some data
* the API for accessing that data

So if you want the newest data, you have to take the newest API (as long as the API is "import a library").

I'm ok with a mix like this:

```text
* breaking code changes           = major version bump
* data-only changes                   = patch version bump
* non-breaking code changes   = minor/patch version bump
```

But if pre-commit-hooks functioning the way we intend depends on `rapids-metadata` following that process, then that should be documented in that repo. The best place would probably be a comment in `pyproject.toml` right about the version number.

That'd help prevent a case like, for example, *"ok it's 2026, I'm going to drop some 24.x versions from this to reduce the package data. I guess that's 'breaking' and should be a major-version bump."*

Reply:
> data-only changes = patch version bump

Yes, this is exactly what I intended. I had not thought of what happens when we remove data, but I suppose that should be a patch bump as well.

Of course, given that we are now looking at the possibility of turning this into a web service, that's going to change the whole development workflow anyway.

Reply:
I've published rapids-metadata 0.3.0 and bumped the dependency on it.